### ISSUE TITLE
Description of `k` in `hard` `PseudotimeKernel` is wrong

### ISSUE THREAD

**Issue by Marius1311** [theislab/cellrank#530]
I basically got this wrong when I wrote down the doscring, currently, it says:
"k
    Number of neighbors to keep for each node, regardless of pseudotime.
    This is done to ensure that the graph remains connected. Only used when `threshold_scheme='hard'`."

That is **not** correct. However, since we have have to change this scheme a bit anyways, let's include the docstring change into the next PR.

**Comment by michalk8** [comment]
My observations thus far, use the below function on pancreas:
```python3
def fn1(
    cell_pseudotime: float,
    neigh_pseudotime: np.ndarray,
    neigh_dist: np.ndarray,
    n_neighs: int = 3,
    k: int = 3,
) -> np.ndarray:
    k_thresh = int(np.ceil(int(0.15 * len(neigh_pseudotime))))  # try adaptive k
    # print(k_thresh)
    # k_thresh = 9
    
    future_ixs, past_ixs = [], []
    for i, nt in enumerate(neigh_pseudotime):
        if cell_pseudotime < nt:
            future_ixs.append(i)
        else:
            past_ixs.append(i)
            
    future_ixs, past_ixs = np.array(future_ixs, dtype=int), np.array(past_ixs, dtype=int)
    if len(past_ixs):
        mask_keep = np.argsort(cell_pseudotime - neigh_pseudotime[past_ixs])[:k_thresh]
        past_ixs_keep = past_ixs[mask_keep]
    else:
        past_ixs_keep = past_ixs
    
    biased_dist = np.zeros_like(neigh_dist)  # -1  # alt. 1
    # biased_dist = -neigh_dist.copy()  # alt. 2
    biased_dist[future_ixs] = neigh_dist[future_ixs]
    biased_dist[past_ixs_keep] = neigh_dist[past_ixs_keep]

    return biased_dist
    # return np.exp(biased_dist)  # uncomment for alt. 1/2
```
- the `k` does not really matter (i.e. even if I remove all past cell indices, the projection looks bad, see Fig. 1)
- what really changes when trying alt. 1/2 (i.e. setting connectivities to -1 for past indices or setting them to neg., see Fig 2. for alt. 1)
- Palantir's adaptive `k` is build in normalized multiscale space, but when filtering past cells, they use PT + the adaptive k, see [here](https://github.com/dpeerlab/Palantir/blob/master/src/palantir/core.py#L328) - I think I'm missing some connection between the multiscale space and PT

Fig. 1
![fig1](https://user-images.githubusercontent.com/46717574/111640944-f5c88680-87fc-11eb-8918-9f57eac1341a.png)
Fig. 2
![fig2](https://user-images.githubusercontent.com/46717574/111640951-f8c37700-87fc-11eb-8d36-1c42e0707672.png)

Not sure how to best solve this... But I have a somewhat related question:
- in scanpy, density normalization happens using the connectivities, but in out code, when density normalizing the biased conn., we still use the old connectivities to build the matrix `Q` - is this correct?

**Comment by Marius1311** [comment]
Hi Mike, let's discuss this in our call today - a couple of things are not really clear to me. For example, why do we suddenly use distances here and not connectivities? 

But I think I can already give you an answer to your last question. I think density normalization should always happen in a symmetric graph because there's nothing directional about a density, that's why I choose to always use the "old", unbiased connectivities for density normalization. If I remember correctly, it didn't make a large difference when I turned density normalization off.

**Comment by michalk8** [comment]
TODOs:
- [ ] use PT to prune past edges, not connectivities
- [ ] use different adaptive `k` @Marius1311 I think you wanted to have a look at this 
- [ ] change default to soft
- [ ] fix the docstring

**Comment by Marius1311** [comment]
Hi Mike, yes, you're right, I will look into the density-dependent k choice!

**Comment by michalk8** [comment]
closed via #536
